Parallel routing

pritamd47 opened the pull request:
Running the routing model in parallel using dask.

SanchitMinocha commented on src/rat/core/run_routing.py:
```diff
@@ -3,8 +3,14 @@
+from pathlib import Path
+import dask
+from tempfile import TemporaryDirectory
+from pathlib import Path
```
Added twice

SanchitMinocha commented on src/rat/core/run_routing.py:
```diff
@@ -30,9 +36,13 @@ def __init__(self, project_dir, result_dir, inflow_dir, model_path, param_path,
 
         self.fdr = self._read_fdr(fdr_path)
-        if os.path.exists(station_path_latlon):
+        if isinstance(station_path_latlon, pd.DataFrame) or isinstance(station_path_latlon, pd.Series):
```
station_path_latlon can not be an instance of pd.series.

SanchitMinocha commented on src/rat/core/run_routing.py:
```diff
@@ -47,90 +57,199 @@ def _get_xy(self, lat, lon) -> tuple:
+        else: # else change directory to project directory
+            args = f'{self.model_path} {self.param_path}'
+            log.debug("Running: %s", " ".join(args))
+            ret_code = subprocess.run(args.split(), capture_output=True, shell=True, cwd=self.project_dir)
```
Add parameter executable="/bin/bash"

It ensures its applicability to all Linux systems even when by default bash shell is not used.

SanchitMinocha commented on src/rat/rat_basin.py:
```diff
@@ -495,43 +497,47 @@ def rat_basin(config, rat_logger):
 
     ######### Step-9
     if(9 in steps):
         try:
-            rat_logger.info("Starting Step-9: Preparation of parameter files for Surface Area Calculation")
+            rat_logger.info("Starting Step-9: Generating Inflow files")
+            #-------------  Inflow pre-processing for Mass Balance begins--------------#
+            ###### Converting routing model's outputs to .csv file
+            routing_output_dir = Path(config['GLOBAL']['data_dir']).joinpath(config['BASIN']['region_name'], 'basins', basin_name, 'ro','ou')
+            dst_dir = Path(config['GLOBAL']['data_dir']).joinpath(config['BASIN']['region_name'], 'basins', basin_name, 'ro', 'rout_inflow')
+            dst_dir.mkdir(parents=True, exist_ok=True)
+            generate_inflow(routing_output_dir, dst_dir)
+        except:
+            no_errors = no_errors+1
+            rat_logger.exception("Error Executing Step-9: Generating Inflow files")
+        else:
+            rat_logger.info("Finished Step-9: Generating Inflow files")
+            #------------- Inflow pre-processing for Mass Balance ends--------------#
```
Merge this with previous Step 13 (currently step 14) so that outflow, area, evaporation and inflow gets calculated in a single step.

SanchitMinocha commented on src/rat/rat_basin.py:
```diff
@@ -495,43 +497,47 @@ def rat_basin(config, rat_logger):
+            routing_output_dir = Path(config['GLOBAL']['data_dir']).joinpath(config['BASIN']['region_name'], 'basins', basin_name, 'ro','ou')
+            dst_dir = Path(config['GLOBAL']['data_dir']).joinpath(config['BASIN']['region_name'], 'basins', basin_name, 'ro', 'rout_inflow')
+            dst_dir.mkdir(parents=True, exist_ok=True)
+            generate_inflow(routing_output_dir, dst_dir)
```
Since now generating inflow is no longer a part of routing, change its destination directory to 'rat outputs' so that it's in line with other rat outputs like sarea, evaporation, outflow. 
Note: You will also have to change the inflow path in Step 13 where inflow is used for calculating outflow and in Step 14 where this data is converted to time series and saved in final outputs.

SanchitMinocha reviewed (changes requested):
Please go through the changes and see if they can be implemented.

pritamd47 commented:
@SanchitMinocha I have implemented the changes requested in the review. This PR also handles the logging issue arising due to dask #40 .

SanchitMinocha commented on src/rat/utils/logging.py:
Hi Pritam, Can you please tell me why this change has been done? Also, is this addressing the sorting of log messages or is dask doing it by self?

SanchitMinocha reviewed (changes requested):
Just was confused with the changes made in logging.py!

SanchitMinocha commented on src/rat/rat_basin.py:
```diff
@@ -220,6 +221,9 @@ def rat_basin(config, rat_logger):
             basin_station_latlon_file = config['ROUTING']['station_latlon_path']
         # Creating routing inflow directory
         rout_inflow_dir = create_directory(os.path.join(basin_data_dir,'ro', 'rout_inflow',''), True)
```
I think you should remove 'rout_inflow_dir' as it won't be used now, I guess.

pritamd47 commented:
I added a formatter which adds information about dask-worker to the log file if called by a dask-worker. If not called by a dask-worker, it prints in the regular style of logging.

I also added a LOG_LEVEL global variable, similar to LOG_NAME. For some reason dask workers seemed to only print warning logs to the file. Setting the LOG_LEVEL = "DEBUG" solves this problem.

SanchitMinocha reviewed (changes requested):
A minor change to remove the creation of a routing inflow directory and making sure it is not anywhere being used.

SanchitMinocha reviewed (approved):
Review of Parallel Routing is complete. Looks good. Well done (y)
